oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/logout/OidcBackChannelLogoutAuthenticationManager.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2002-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.client.oidc.authentication.logout;
+
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.authentication.AuthenticationServiceException;
+import org.springframework.security.authentication.BadCredentialsException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.jwt.BadJwtException;
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.JwtDecoderFactory;
+import org.springframework.security.oauth2.jwt.JwtException;
+
+public class OidcBackChannelLogoutAuthenticationManager implements AuthenticationManager {
+
+	private final JwtDecoderFactory<ClientRegistration> logoutTokenDecoderFactory = new OidcLogoutTokenDecoderFactory();
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		if (!(authentication instanceof OidcBackChannelLogoutAuthenticationToken token)) {
+			return null;
+		}
+		String logoutToken = token.getLogoutToken();
+		ClientRegistration registration = token.getClientRegistration();
+		JwtDecoder logoutTokenDecoder = this.logoutTokenDecoderFactory.createDecoder(registration);
+		try {
+			return new OidcBackChannelLogoutAuthentication(OidcLogoutToken.withTokenValue(logoutToken)
```

Comment:
This isn't a place where Spring Security is consistent. Most recently, `Saml2AuthenticationToken` and `Saml2Authentication` were constructed with the express purpose of separating authentication request and result instances. 

I'll circle back with the team on this one.

Reply:
The specific difficulty here is that the principal type is different. The authentication token takes a `String` and the authentication result takes an `OidcLogoutToken`.

Reply:
It's needed to have two since the first has a `String` principal and the second has an `OidcLogoutToken` principal. This allows the intent of the resulting authentication to be clear as well as being more type-safe.

Case in point, please see how this simplifies `OidcBackChannelLogoutHandler` in the latest version of the PR to not have to cast in order to get an `OidcLogoutToken`.

Reply:
This is a separate but related discussion that will likely resolve this comment. After looking at `OidcBackChannelLogoutHandler`,  I realized that it's not actually a `LogoutHandler` as it does not perform any logout operations. It simply "requests" that the logout be performed, which is actually performed by the `LogoutHandler` associated to the `LogoutFilter`. Having said that, I feel that we should push the logic contained in `OidcBackChannelLogoutHandler` into `OidcBackChannelLogoutAuthenticationProvider` and remove `OidcBackChannelLogoutHandler`. What are your thoughts?

Reply:
I disagree. The authentication provider should just authenticate.

While there are times when an authentication provider does operate on state (like the `state` parameter), it is only in the service of authenticating the request. 

> I realized that it's not actually a LogoutHandler as it does not perform any logout operations

Operating on the registry and making a request to logout is performing a logging out. Just because an authentication provider queries an external entity to authenticate, it doesn't mean it's not authenticating (consider `RemoteAuthenticationProvider` (now defunct) or `OpaqueTokenAuthenticationProvider` which requests that the token be authenticated). The same logic holds with logging out.

Reply:
> The authentication provider should just authenticate.

Why should it be limited to just authenticating? Is there an issue with performing subsequent operations after authenticating the request? FYI, there are current implementations of `AuthenticationProvider` that perform subsequent operations after authenticating the request and modifying `OidcSessionRegistry` (with removal) would be a candidate.

Do you foresee any issue with pushing the logic in `OidcBackChannelLogoutHandler` into `OidcBackChannelLogoutAuthenticationProvider`.

Reply:
The information from the session registry is needed to be able to perform the logout -- as far as I can tell, this means we'd either 

1. Violate transaction boundaries in the event of a datasource-based implementation; the transaction would need to be started inside the authentication provider and then preserved until the logout handler (to leave the `Iterable` uniterated). This feels like [the open-session-in-view anti-pattern](https://vladmihalcea.com/the-open-session-in-view-anti-pattern/) to me. Or,
2. Change the call in authentication provider to get-then-remove, which creates the following issues:
a. The API is no longer minimal (we added a method that's unnecessary in the current proposal)
b. It's needed to configure `OidcSessionRegistry` in two public logout APIs now in order to be correct
c. The `Iterable` is no longer effective in keeping things out of memory unnecessarily (unless we go back to 1)
d. The removal, by definition, is no longer atomic

What is the issue with the proposed arrangement?

Reply:
After looking at the current `OidcBackChannelLogoutHandler` again, I also feel that the component should be removed in favor of pushing its logic into the `AuthenticationProvider`.

Currently, this PR doesn't do much with the response in a success scenario, but this could be added later and would be part of the filter's responsibility (probably through an `AuthenticationSuccessHandler`). I don't think the `LogoutHandler` should directly render a response. So the issue I see with this logic residing in a `LogoutHandler` is that it is more difficult for an implementation to collaborate with the filter to influence the response since it can't return anything.

More generally, I feel like the conversation for Spring Security 7 would involve discussing and solving for a general pattern around endpoints that perform work like this. I don't personally feel it's the best choice, but `AuthenticationProvider` is what we have at the moment. For that reason I don't see an issue pushing this logic into the `AuthenticationProvider` for now.

Reply:
@jzheaux There might be some confusion from my last comment based on your comment:

> Violate transaction boundaries in the event of a datasource-based implementation; the transaction would need to be started inside the authentication provider and then preserved until the logout handler (to leave the Iterable uniterated). This feels like [the open-session-in-view anti-pattern](https://vladmihalcea.com/the-open-session-in-view-anti-pattern/) to me

**I'm not proposing** to reuse `OidcBackChannelLogoutHandler` inside `OidcBackChannelLogoutAuthenticationProvider`.

Instead, I'm proposing we move the logic in `OidcBackChannelLogoutHandler.logout()` to `OidcBackChannelLogoutAuthenticationProvider.authenticate()` and remove `OidcBackChannelLogoutHandler`. This way all the authentication logic and subsequent operation of sending the logout request is performed inside `OidcBackChannelLogoutAuthenticationProvider`. This maintains the transaction boundary within `OidcBackChannelLogoutAuthenticationProvider` (the "service").

> What is the technical issue with the proposed arrangement?

The strategy used to implement the logout using the "self-call" is a unique implementation IMO and I would rather hide that logic for now. With the current exposed implementation of `OidcBackChannelLogoutHandler`, it further exposes `setRestOperations()`, which is implementation specific and my preference would be to completely hide the implementation.

Reply:
I agree that having just one component would address the concerns I raised [here](https://github.com/spring-projects/spring-security/pull/12570#discussion_r1293646364). Thank you for clarifying. I did not consider the possibility of placing all logic into a single `AuthenticationProvider`.

Before getting to that, let me mention some points that feel some agreement on:

> it further exposes `setRestOperations()`
> -- Joe

I think it's fine to remove `setRestOperations` for the time being. I see your point about wanting to give that implementation room to evolve. If this is your only technical concern, then I say we remove the method and proceed with the current arrangement.

> I don't think the `LogoutHandler` should directly render a response. 
> -- Steve

This seems likely. It does feel odd to me that `LogoutHandler` would attempt to affect, say, the status code of the filter's response. In light of that, I'm open to changing to `AuthenticationSuccessHandler`.

Regarding the counter-proposal, I don't think that placing all business logic in an `AuthenticationProvider` is the best arrangement as it violates its JavaDoc, causes the PR to send request-specific information to the service layer, and departs from how other OAuth 2.0 filters separate authentication from session management. Allow me to expand on each:

First, doing so dramatically widens the scope of `AuthenticationManager`'s JavaDoc which says:

```java
/**
  * Attempts to authenticate the passed Authentication object, returning a fully populated
  * Authentication object (including granted authorities) if successful.
  **/ 
```

According to this JavaDoc, now in my own words, to `authenticate` is to determine if the given material is authentic and return a fully-populated principal (`Authentication` extends `Principal`).

What role does mutating the `OidcSessionRegistry` have in determining the logout token to be authentic? (To *lookup* the  sessions is arguably part of the "fully populated principal", as already mentioned in my [option 2](https://github.com/spring-projects/spring-security/pull/12570#discussion_r1293646364), but not to mutate them.)

Second, logging out will require the `HttpServletRequest` because the logout invocation needs to look up the base URL from the current request to correctly formulate the internal `/logout` request. I don't think that we want to add something like `baseUrl` to the `OidcLogoutAuthenticationToken` implementation. While possible, the veneer as a service layer becomes quite thin at that point.

Third, this is not how other components interact with `SessionRegistry` as a possible parallel. For example, `OAuth2LoginAuthenticationFilter` invokes `RegisterSessionAuthenticationStrategy` after `AuthenticationManager` returns. IOW, a component separate from `AuthenticationManager` is what operates on the registry. While I want to find the right solution for the logout filter, I don't think we should dismiss how the other OAuth 2.0 filters separate authentication from session management.

Reply:
> In light of that, I'm open to changing to AuthenticationSuccessHandler.

This could be a good compromise.

Reply:
@jgrandja, @sjohnr, and I talked about this offline. 

For now, I'm going to try and make the API private, just exposed to the DSL, to give the team more time to get on the same page while still allowing the feature to be released. I've created #13767 as a follow-up.